### ISSUE TITLE
[HOLD for payment 2023-08-21] [$250] mWeb - Unresponsive view on 2 factors authentication screen

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#23481]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to profile settings
2. Then click on security
3. Click on two factors authentication
4. Click on copy and and then go to next
5. Click on Textfield
6. Observe the screen

## Expected Result:
 Ui should be responsive. And show properly even keyboard is focused.

## Actual Result:
Screen is unresponsive. Some view disappears while focused.

## Workaround:
Unknown 

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [x] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.44-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/a1d0877a-62f3-41d3-ad09-ad3147b853cd

https://github.com/Expensify/App/assets/93399543/e9361ff2-44d4-4229-8e6b-9edeb069e574

**Expensify/Expensify Issue URL:**
**Issue reported by:** @niravkakadiya25
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1690115036631379

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01deea2e8f64727cde</li>
        <li>Upwork Job ID: 1683466421789040640</li>
        <li>Last Price Increase: 2023-08-08</li>
<li>Automatic offers: </li>
<ul>
<li>s77rt | Reviewer | 26018994</li>
<li>jscardona12 | Contributor | 26018995</li>
<li>niravkakadiya25 | Reporter | 26018997</li>
</ul></li>
    </ul>
</details>

**Comment by trjExpensify** [comment]
Yeah, I can repro this one. The `next` confirmation button should be bottom-docked on this page, like the others. 👍

**Comment by ShogunFire** [comment]
It was happening when we had max height on keyboardAvoidingView  but was supposed to be fixed here: https://github.com/Expensify/App/pull/16443/files

**Comment by DevWizard0000** [comment]
### Proposal
### Please re-state the problem that we are trying to solve in this issue.
Responsive view when the keyboard appears.

### What is the root cause of that problem?
Cover the full page with KeyAvoidingView or SafeAreaView.

### What changes do you think we should make in order to solve the problem?
Cover the full page with KeyAvoidingView or SafeAreaView.
And re-style the Next Button.

**Comment by s77rt** [comment]
@samh-nl Thanks for the proposal. I don't think your RCA is correct.

**Comment by s77rt** [comment]
@ShogunFire Yeah I remember such a bug. This one is probably a regression. This works fine on version `1.3.43-0`.

**Comment by s77rt** [comment]
@DevWizard0000 Thanks for the proposal. Please follow the proposal template, we need to find the root cause first then we will work out a solution.

**Comment by s77rt** [comment]
Not overdue. Still looking for proposals

**Comment by trjExpensify** [comment]
@DevWizard0000 are you interesting in revising your proposal into the correct format?

**Comment by studentofcoding** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Unresponsive view on 2 factors authentication screen on Mobile

### What is the root cause of that problem?
[App/src/styles/getNavigationModalCardStyles/index.website.js](https://github.com/Expensify/App/blob/8b75ed647769a96bc68e740fd21b4c80e6c4434e/src/styles/getNavigationModalCardStyles/index.website.js#L13)

Line 13 in [8b75ed6](https://github.com/Expensify/App/commit/8b75ed647769a96bc68e740fd21b4c80e6c4434e)

Use KeyAvoidingView or SafeAreaView to cover the full page

### What changes do you think we should make in order to solve the problem?
Cover the full page with KeyboardAvoidingView or SafeAreaView.

### What alternative solutions did you explore? (Optional)
N/A

**Comment by jscardona12** [comment]
@ginsuma @s77rt it is needed because with the new navigation implementation the minHeight attribute was on develop. The fix shouldn't be deleting it, it should be setting up to 0.

**Comment by s77rt** [comment]
@studentofcoding Thanks for the proposal. You mentioned the offending PR but there is no root cause analysis. Can you please explain why that PR caused the regression?

**Comment by jscardona12** [comment]
## Proposal
Please re-state the problem that we are trying to solve in this issue.
Unresponsive view on 2 factors authentication screen on Mobile

### What is the root cause of that problem?
The root cause is that because of 
[App/src/styles/getNavigationModalCardStyles/index.website.js](https://github.com/Expensify/App/blob/8b75ed647769a96bc68e740fd21b4c80e6c4434e/src/styles/getNavigationModalCardStyles/index.website.js#L13) 
the height of the settings page will always be the window height. 

The Root cause is that every time the keyboard appears on safari iit scrolls up the window changing the window height to windowHeight - window.scrollY. When the height become smaller the Next button goes up and do not stay in position because the styles are based on flex.

The minHeight is needed because it fixes the 3 mentioned issues in the code comments.


### What changes do you think we should make in order to solve the problem?
We need to take add to the minHeight the window.scrollY. if the keyboard is not open it is 0. if it's open it would add the height that is being taken off the settingsInitialPage. 

        export default ({windowHeight, isSmallScreenWidth}) => ({
        ...getBaseNavigationModalCardStyles({isSmallScreenWidth}),

    // This height is passed from JavaScript, instead of using CSS expressions like "100%" or "100vh", to work around
    // Safari issues:
    // https://github.com/Expensify/App/issues/12005
    // https://github.com/Expensify/App/issues/17824
    // https://github.com/Expensify/App/issues/20709

    height: windowHeight + window.scrollY,
    minHeight: windowHeight,
  });

Solution Working: 
 

https://github.com/Expensify/App/assets/16781411/465480aa-d297-40da-80bc-d4f9a5e7a3da

**Comment by s77rt** [comment]
@jscardona12 Thanks for the proposal. Can you explain why this issue only occurs in Safari?

**Comment by jscardona12** [comment]
@s77rt it's a safari known issue.

https://bugs.webkit.org/show_bug.cgi?id=141832
Safari it's the only browser that considers this a feature. 

![image](https://github.com/Expensify/App/assets/16781411/f8f5a5b4-392e-40cd-9e2d-539a4d54ce0e)

Here is an article explaining it better: https://blog.opendigerati.com/the-eccentric-ways-of-ios-safari-with-the-keyboard-b5aa3f34228d

**Comment by jscardona12** [comment]
@mvtglobally  it's steal reproducible on Dev

**Comment by s77rt** [comment]
@jscardona12 Thanks for the update. I think having the height controlled by the scroll position is a hacky solution. Even the very first solutions (https://github.com/Expensify/App/pull/12509 and https://github.com/Expensify/App/pull/19682) looks hacky, it may be better to rethink an entire new approach. Safari issues are tricky to handle and they mostly backfire.

**Comment by s77rt** [comment]
@jscardona12 Thanks for the update. I don't think the solution is reliable, and we don't want to disable a feature to fix a bug.

**Comment by jscardona12** [comment]
@s77rt the solution wouldn't be reliable because of the disabling the keyboard autofocus or in general? If in general, why would it be unreliable?

**Comment by s77rt** [comment]
@jscardona12 I didn't test the solution but the fact that it does not work in all the cases e.g. in autofocus makes it unreliable.

**Comment by trjExpensify** [comment]
Ah interesting, I wonder whether the breakthrough from @staszekscp in https://github.com/Expensify/App/pull/23922 will help here?

**Comment by s77rt** [comment]
@trjExpensify I think that issue is unrelated to this

**Comment by trjExpensify** [comment]
Fair enough!

**Comment by s77rt** [comment]
Does not look we are close Melvin. Safari issues are tricky issues.

**Comment by trjExpensify** [comment]
Bumping the price to $2k! 

@s77rt, out of curiosity, how are we're achieving this on other screens?
![IMG_0141](https://github.com/Expensify/App/assets/16232057/30dc813d-0f04-4211-a5e1-e73102a3f7b4)

**Comment by s77rt** [comment]
@trjExpensify The bug only occurs if you focus an input that is located in the bottom half of the screen (or located in where the keyboard would normally show up) in that case Safari will scroll up which does not seem to work well with the solution added in https://github.com/Expensify/App/pull/23222.

**Comment by s77rt** [comment]
@jscardona12 Thanks for the update. I think having to add `position:'absolute', height: '100%'` to every effected page (e.g. `VerifyPage`) is a workaround. Can we have a solution that makes the changes to the `ScreenWrapper` itself? (global solution)



Note: Please do not write every new proposal in a new comment, instead just update the existing one.

**Comment by jscardona12** [comment]
@s77rt in the updated version of main, the keyboard is not getting openes with autofocus, you need to tap the input for it to appear. i dont know if this is an error or an expected behaviour (@trjExpensify) If this is an expected behaviour we dont need to do anything on ScrapperWrapper if it is? should this bug be reported?

**Comment by s77rt** [comment]
@jscardona12 It's still auto focusing from my side.

**Comment by jscardona12** [comment]
@s77rt that's weird, in my phone the input is focused but the keyboard does not appeared.

https://github.com/Expensify/App/assets/16781411/3b1bce60-0a28-4218-9878-eaae96aaf245

**Comment by s77rt** [comment]
@jscardona12 That's another issue I think. Safari limits the auto focus functionality. This works fine on the Simulator though.

**Comment by jscardona12** [comment]
@s77rt  But it worked before i made merge, should we base behaviour on real devices or in simulators?

**Comment by s77rt** [comment]
@jscardona12 Let make sure it works on both (the auto focus is another problem).

**Comment by jscardona12** [comment]
@s77rt proposal updated. In the simulator and in real device is working fine.

**Comment by s77rt** [comment]
@jscardona12 Can you please revisit your proposal? It does not seem right? I can't read it :sweat_smile: Also it seems you are suggesting to remove `height` and and `minHeight`, please include that in your proposal as well.

**Comment by s77rt** [comment]
@trjExpensify Can you please update the title here? There should be only $250 to pay to the bug reporter. The rest of the payment will be handled in the original issue (since this is a regression).

**Comment by jscardona12** [comment]
@s77rt updated.

**Comment by trjExpensify** [comment]
>@trjExpensify Can you please update the title here? There should be only $250 to pay to the bug reporter. The rest of the payment will be handled in the original issue (since this is a regression).

Ohh, I totally missed that was a recent PR of @jscardona12's. However, shall we remove `help wanted` then as well and get this issue assigned? It's currently open to proposals versus an assigned issue to fix a recent regression introduced by the PR author.

**Comment by s77rt** [comment]
@trjExpensify Ah right, please assign @jscardona12 and remove the `Help Wanted` label.

**Comment by s77rt** [comment]
@jscardona12 Thanks for the update. The solution looks good to me but I'm still trying to reproduce the original bug :sweat_smile: Hopefully we should be good to go soon

**Comment by s77rt** [comment]
I chatted with @jscardona12 1:1 and it looks we only needed `position: fixed` to fix https://github.com/Expensify/App/issues/20709 (and https://github.com/Expensify/App/issues/17824), we can revert the other changes that caused this regression.

While checking https://github.com/Expensify/App/issues/17824 I found that @hellohublot already [proposed](https://github.com/Expensify/App/issues/17824#issuecomment-1528704182) the `position: fixed` solution. @hellohublot Is there any specific reason why we should not go with that solution? cc @eVoloshchak @cubuspl42

**Comment by jscardona12** [comment]
@s77rt @trjExpensify [PR](https://github.com/Expensify/App/pull/24305) ready

**Comment by cubuspl42** [comment]
@s77rt If this works for all possible cases, go for it. Unfortunately, my intuition with iOS Safari viewport problems so far is that it's unfixable, and when you hit one nail to work around one of its bugs, things break in another place. I'm sorry that it seems that I joined the `(fix, regression)*` loop.

**Comment by s77rt** [comment]
- The PR that introduced the bug has been identified: https://github.com/Expensify/App/pull/19682 and https://github.com/Expensify/App/pull/23222
- The offending PR has been commented on: https://github.com/Expensify/App/pull/19682#discussion_r1294614414 and https://github.com/Expensify/App/pull/23222#discussion_r1294614456
- A discussion in #expensify-bugs has been started: Not needed. This is a regression that could have been caught during PR testing
- Determine if we should create a regression test for this bug: No

**Comment by trjExpensify** [comment]
👋 payments are due here! As a regression, confirming that's $250 for the bug report only correct @s77rt?

**Comment by trjExpensify** [comment]
Cool, thanks for confirming. Settled up with @niravkakadiya25. Closing!

**Comment by sophiepintoraetz** [comment]
@trjExpensify @s77rt @jscardona12 - callstack [suggested](https://expensify.slack.com/archives/C03UK30EA1Z/p1693469360430019?thread_ts=1693429807.053229&cid=C03UK30EA1Z) the PR caused [this regression](https://github.com/Expensify/App/issues/24397) - mind taking a look and confirming? If so, I'll hand it over to the contributors to fix.

**Comment by s77rt** [comment]
@sophiepintoraetz Probably unrelated. That issue was reported to be reproducible in 1.3.52 but the PR targeting this issue was only deployed in 1.3.53.

FWIW I can't reproduce the linked issue.

**Comment by trjExpensify** [comment]
Hm, not sure if they are related on a first look, but @s77rt @cubuspl42 @jscardona12 might have a better idea!

**Comment by trjExpensify** [comment]
^^ shoulda' refreshed. 😂

**Comment by trjExpensify** [comment]
Cool, @sophiepintoraetz I'm re-closing this done issue and will leave that new bug report up to yah to deal with.

**Comment by sophiepintoraetz** [comment]
Thanks @trjExpensify and @s77rt for the confirmation!
